tokio/src/runtime/task/inject.rs

```diff
@@ -107,34 +113,38 @@ impl<T: 'static> Inject<T> {
     }
 
     pub(crate) fn pop(&self) -> Option<task::Notified<T>> {
+        self.pop_n(1).next()
+    }
+
+    pub(crate) fn pop_n(&self, n: usize) -> Pop<'_, T> {
+        use std::cmp;
+
         // Fast path, if len == 0, then there are no values
         if self.is_empty() {
-            return None;
+            return Pop {
+                len: 0,
+                pointers: None,
+                _p: PhantomData,
+            };
         }
 
-        let mut p = self.pointers.lock();
-
-        // It is possible to hit null here if another thread popped the last
-        // task between us checking `len` and acquiring the lock.
-        let task = p.head?;
-
-        p.head = get_next(task);
-
-        if p.head.is_none() {
-            p.tail = None;
-        }
+        // Lock the queue
+        let p = self.pointers.lock();
 
-        set_next(task, None);
+        let n = cmp::min(n, unsafe { self.len.unsync_load() });
 
         // Decrement the count.
         //
         // safety: All updates to the len atomic are guarded by the mutex. As
         // such, a non-atomic load followed by a store is safe.
         self.len
-            .store(unsafe { self.len.unsync_load() } - 1, Release);
+            .store(unsafe { self.len.unsync_load() } - n, Release);
```

Comment:
Maybe only `unsync_load` once? That also fixes the missing safety comment on the first one.

Reply:
Done